### DISCUSSION TITLE
How to create a vtkPolyData object from pyvista mesh?

### DISCUSSION THREAD

**Discussion by gofmans** [pyvista/pyvista#3666] [Q&A] [answered]
Hello all,

I want to use a specific filter of raw vtk and for this, I need to convert my pyvista mesh to vtkPolyData object.

I kinda found a way to do that but it's SLOW and requires iterating over all triangles of the mesh, and I'm sure there is a simpler and more efficient way to do that.

Here is my current version:

```python
from vtk.util.numpy_support import numpy_to_vtk
import pyvista as pv
import vtk

# load mesh
my_pyvista_mesh = pv.read(**path to specific stl file**)

# reshape faces to format: (3, x, y, z)
faces = my_pyvista_mesh.faces.reshape((int(len(my_pyvista_mesh.faces)/4)),4)

# create vtkPoints object from mesh points
numberPoints = len(my_pyvista_mesh.points)
Points = vtk.vtkPoints()
points_vtk = numpy_to_vtk(np.asarray(my_pyvista_mesh.points, order='C'), deep=1)
Points.SetNumberOfPoints(numberPoints)
Points.SetData(points_vtk)


# create vtkCellArray object from mesh faces
Triangles = vtk.vtkCellArray()
for face in faces:
    Triangle = vtk.vtkTriangle()
    Triangle.GetPointIds().SetId(0,face[1])
    Triangle.GetPointIds().SetId(1,face[2])
    Triangle.GetPointIds().SetId(2,face[3])
    Triangles.InsertNextCell(Triangle)

# create the desired vtkPoltData object from points and faces
polydata = vtk.vtkPolyData()
polydata.SetPoints(Points)
polydata.SetPolys(Triangles)
```

Any help would be greatly appreciated :)

Keep up the great work!

**Comment by MatthewFlamm** [comment]
Can you elaborate further on why you cannot use the pyvista object directly?  The pyvista object inherits from `vtkPolyData` so it can generally be used as one.  In fact, this is how all the internal filters work, we apply vtk filters to pyvista objects.

**Comment by gofmans** [comment, marked as answer]
Thanks, @MatthewFlamm for the tip!

I tried loading the mesh directly to the filter and all I had to do is use the "SetInputData()" option instead the SetInputConnection():

```
# load mesh
my_pyvista_mesh = pv.read(**path to specific stl file**)

specific_vtk_filter = vtk.vtkFillHolesFilter()

# this works
specific_vtk_filter.SetInputData(my_pyvista_mesh)

# this fails
specific_vtk_filter.SetInputConnection(my_pyvista_mesh)
```
